### PR TITLE
Complexity adjusted scoring

### PR DESCRIPTION
**Pull request by Audreym98**
This PR adds an argument that allows alignments to be scored using complexity adjusted scoring. 

Updates to the converter have also been made to include the alignment tool in the sto file and background frequencies in the matrix files, both are used to determine if complexity adjusted scoring can be done. 

Additionally, the cm_to_stockholm converter is set up to be called by an alignment_to_stockhom script which takes an alignment file and tool to run the correct converter. Other converter options will need to be added to this script for future use.

### CODE REVIEW

File: converters/rm_to_stockholm.py [hunk 1]
```diff
@@ -136,7 +146,7 @@ def print_alignment(chrom_seq, subfam_seq, chrom, subfam, f_out_sto):
     f_out_sto.write("//\n")
 
 
-def print_score_matrix(filename_out_matrix, score_matrix, matrix_name):
+def print_score_matrix(f_out_matrix, score_matrix, matrix_name):
```
Comment (glesica) [inline converters/rm_to_stockholm.py:149]: Good catch!

File: polyA/__init__.py [hunk 2]
```diff
@@ -4,4 +4,72 @@
...
+    load_substitution_matrices,
+    SubMatrix,
+    SubMatrixCollection,
+)
+from .ultra_provider import (
+    UltraProvider,
+    UltraOutput,
+    ApplicationUltraProvider,
+    TandemRepeat,
+)
```
Comment (glesica) [inline polyA/__init__.py:74]: Why did all of this get re-added? Since PolyA the tool is really the only consumer of PolyA the library, I had removed all the re-exports that were in this file as they are redundant.

File: polyA/_app.py [hunk 3]
```diff
@@ -94,10 +95,23 @@ def run():
     # Load the substitution matrix
     # ----------------------------
 
+    with open(opts.alignments_file_path) as _infile:
+        with open(opts.alignments_file_path) as _infile:
```
Comment (glesica) [inline polyA/_app.py:99]: Why open it twice, and why shadow the `_infile` name?

File: polyA/_app.py [hunk 4]
```diff
@@ -94,10 +95,23 @@ def run():
...
+            alignment_tool: str = load_alignment_tool(_infile)
+
+    # Note: alignments from blast or HMMER are not set-up to use complexity adjusted scoring
+    if opts.complexity_adjustment and alignment_tool not in [
+        "cross_match",
+        "RepeatMasker",
+    ]:
+        raise AppError(
+            "cannot use complexity adjusted scoring with this alignment tool"
```
Comment (glesica) [inline polyA/_app.py:108]: Add the alignment tool name to the error message:

```
f"cannot use complexity adjusted scoring with {alignment_tool}"
```

File: polyA/calculate_score.py [hunk 5]
```diff
@@ -9,6 +10,7 @@ def calculate_score(
     prev_char_seq1: str,
     prev_char_seq2: str,
     sub_matrix: Dict[str, int],
+    char_complexity_adjustment: Dict[str, int] = None,
```
Comment (glesica) [inline polyA/calculate_score.py:13]: The type here should be `Optional[Dict[str, int]]` since you're setting `None` as the default value.

File: polyA/calculate_score.py [hunk 6]
```diff
@@ -70,5 +74,81 @@ def calculate_score(
...
-
+            if char_complexity_adjustment:
+                chunk_score += char_complexity_adjustment[seq2_char]
     return chunk_score
+
+
+def calc_target_char_counts(
+    query_seq: str,
+    target_seq: str,
+):
```
Comment (glesica) [inline polyA/calculate_score.py:85]: Add a doc comment that explains what this does at a high level. In other words, how does this function fit into the overall algorithm, and why would someone want to call it?

File: polyA/calculate_score.py [hunk 7]
```diff
@@ -70,5 +74,81 @@ def calculate_score(
...
+    total_chars = 0
+    for q_base, t_base in zip(query_seq, target_seq):
+        if t_base not in ["A", "G", "T", "C"]:
+            other_chars.add(t_base)
+            continue
+        if q_base == "-":
+            continue
+        query_char_counts[t_base] += 1
+        total_chars += 1
+    return query_char_counts, total_chars, other_chars
```
Comment (glesica) [inline polyA/calculate_score.py:97]: `total_chars` is easy to compute from `query_char_counts` so we probably don't need to return it. You can get it with `sum(query_char_counts.values())`.

File: polyA/calculate_score.py [hunk 8]
```diff
@@ -70,5 +74,81 @@ def calculate_score(
...
+    char_background_freqs: Optional[Dict[str, float]],
+    query_seq: str,
+    target_seq: str,
+    lamb: float,
+):
+    char_complexity_adjustments: Dict[str, float] = {}
+    if char_background_freqs is None:
+        # set the complexity adjustment to zero for every char
+        for char in target_seq:
+            char_complexity_adjustments[char] = 0
```
Comment (glesica) [inline polyA/calculate_score.py:110]: Small issue, but it's better to initialize a float with `0.0`. If you do that, you also don't need the type annotation on the variables below, so instead of `t_factor: float = 0` you can just do `t_factor = 0.0` and the type checker will infer the correct type.

File: polyA/calculate_score.py [hunk 9]
```diff
@@ -70,5 +74,81 @@ def calculate_score(
...
+    t_sum: float = 0
+    t_counts: int = 0
+    target_char_counts, total_chars, other_chars = calc_target_char_counts(
+        query_seq, target_seq
+    )
+
+    char_complexity_adjustments = {"A": 0, "C": 0, "G": 0, "T": 0}
+    for char, freq in char_background_freqs.items():
+        count = target_char_counts[char]
+        if count > 0 and freq > 0 and math.log(freq) != 0:
```
Comment (glesica) [inline polyA/calculate_score.py:125]: The log of `freq` is going to be 0 exactly when `freq == 1`, right? So why not test that instead, since it's simpler?

File: polyA/calculate_score.py [hunk 10]
```diff
@@ -70,5 +74,81 @@ def calculate_score(
...
+        total_chars += 1
+    return query_char_counts, total_chars, other_chars
+
+
+def calculate_complexity_adjusted_score(
+    char_background_freqs: Optional[Dict[str, float]],
+    query_seq: str,
+    target_seq: str,
+    lamb: float,
+):
```
Comment (glesica) [inline polyA/calculate_score.py:105]: Add a doc comment that explains this function in context. Also, include the actual computations being done. Pretend it's a math paper and you need to define, precisely, "complexity adjusted score". Feel free to use (pseudo) math notation even. This can help when debugging the function, it can also make it easier to write tests, since the definition is likely to be far more concise than the actual code since it can elide things like error handling.

Comment (glesica) [reply polyA/calculate_score.py:105]: Also add some doc tests to this function and the one above.

File: polyA/calculate_score.py [hunk 11]
```diff
@@ -70,5 +74,81 @@ def calculate_score(
...
+    # char value contribution to t_sum
+    for c, v in char_complexity_adjustments.items():
+        char_complexity_adjustments[c] *= t_sum
+
+    # per position char value contribution to score adjustment
+    for c, v in char_complexity_adjustments.items():
+        if v != 0:
+            char_complexity_adjustments[c] /= target_char_counts[c]
+            char_complexity_adjustments[c] /= lamb
+            char_complexity_adjustments[c] += 0.999 / total_chars
```
Comment (glesica) [inline polyA/calculate_score.py:148]: Where does `0.999` come from? It should probably be a constant in `constants.py`, along with documentation explaining what it is and why it has the value it has.

File: polyA/converters/alignment_to_stockholm.sh [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+  echo "$0: takes 2 arguments: alignment file, alignment tool" 1>&2
+  exit 1
+fi
+
+alignment_file="$1"
+alignment_tool="$2"
+
+if [ "$alignment_tool" = "cross_match" ]; then
+  python cm_to_stockholm.py "$alignment_file"
+fi
```
Comment (glesica) [inline polyA/converters/alignment_to_stockholm.sh:15]: We also have a RM converter, right? The shell script probably ought to live outside the Python package, also.

File: polyA/converters/cm_to_stockholm.py [hunk 13]
```diff
@@ -181,13 +181,17 @@ def print_alignment(chrom_seq, subfam_seq, chrom, subfam, f_out_sto):
...
 
 
-def print_score_matrix(filename_out_matrix, score_matrix, matrix_name):
+def print_score_matrix(
+    filename_out_matrix, score_matrix, matrix_name, background_freqs
+):
     """
     print score matrix to its own output file with extension ".matrix"
     """
     f_out_matrix = open(filename_out_matrix, "w")
```
Comment (glesica) [inline polyA/converters/cm_to_stockholm.py:190]: While we're here, we should change this to use a `with` to open the file, or else close it at the end of the function.

File: polyA/calculate_score.py (hunk 10 above)
Comment (glesica) [reply polyA/calculate_score.py:105]: BTW there's a way to insert LaTeX style math notation into Python docstrings. See https://stackoverflow.com/questions/10081633/using-displaymath-directives-in-docstrings-formulas for an example.

File: polyA/fill_align_matrix.py [hunk 14]
```diff
@@ -62,7 +66,8 @@ def fill_align_matrix(
     >>> subs = ["", "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA...............", "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA--A..............."]
     >>> strts = [0, 2, 0]
     >>> sub_mat = [{"AA":1, "AT":-1, "TA":-1, "TT":1, "..":0}] * 3
-    >>> m = fill_align_matrix([0.0, 0.1, 0.1], 41, 0, 31, [0, -25, -25], [0, -5, -5], 1.0, subs, chros, strts, sub_mat)
+    >>> back_freqs = [None] * 3
+    >>> m = fill_align_matrix([0.0, 0.1, 0.1], 41, 0, 31, [0, -25, -25], [0, -5, -5], 1.0, subs, chros, strts, sub_mat, back_freqs)
```
Comment (glesica) [inline polyA/fill_align_matrix.py:70]: Let's add a regular unit test suite (see `load_alignments_test.py` in `tests/`) so that we can test with "real" background frequencies in addition to `None` values. Let me know if you want help with that.

File: polyA/fill_align_matrix.py [hunk 15]
```diff
@@ -79,6 +84,12 @@ def fill_align_matrix(
         gap_init = gap_inits[i]
         gap_ext = gap_exts[i]
 
+        char_complexity_adjustments: Dict[
+            str, int
+        ] = calculate_complexity_adjusted_score(
```
Comment (glesica) [inline polyA/fill_align_matrix.py:89]: Strictly speaking you don't need the type annotation on the variable as long as the function as a return type annotation. So I would move the type annotation to `calculate_complexity_adjusted_score`. For example:

```
def foo() -> int:
  pass
```

defines a function that returns an `int`.

File: polyA/load_alignments.py [hunk 16]
```diff
@@ -180,6 +180,20 @@ def load_alignments(
...
+    """
+    Return alignment tool used.
+    """
+    for line in file:
+        if _parse_preamble_line(line):
+            continue
+        if line.strip().upper().startswith("# ALIGNMENT TOOL"):
+            # get alignment tool
+            return line.split()[-1]
+        return ""
```
Comment (glesica) [inline polyA/load_alignments.py:193]: Why is this here? It causes the loop to only ever run once, right? If that's the behavior you want, the more standard way to do it would be:

```python
try:
  line = next(file)
except StopIteration:
  return ""

# Do stuff with line
```

File: polyA/substitution_matrix.py [hunk 17]
```diff
@@ -17,9 +17,12 @@ class ParseError(RuntimeError):
 
 class SubMatrix:
     """
-    A single substitution matrix, including its name and lambda value.
+    A single substitution matrix, including its name, lambda value, and assumed
+    character background frequencies.
     The lambda value may be `None`, in which case it must be computed
     before the matrix can be used.
+    The background frequencies may also be 'None' if complexity adjusted scoring
+    will not be used.
```
Comment (glesica) [inline polyA/substitution_matrix.py:25]: Good!

File: polyA/substitution_matrix.py [hunk 18]
```diff
@@ -60,6 +65,23 @@ def _parse_matrix_header(line: str) -> Tuple[str, Optional[float]]:
...
+    each char and it's assumed background frequency.
+    """
+    matrix_background_freqs: Dict[str, float] = {}
+    if line.strip().upper().startswith("BACKGROUND FREQS"):
+        clean_line = re.sub(r"^\s+|\s+$", "", line)
+        if line.strip().upper().startswith("BACKGROUND FREQS"):
+            # convert to dictionary
+            string_dict = clean_line[
+                clean_line.find("{") : clean_line.find("}") + 1
+            ]
```
Comment (glesica) [inline polyA/substitution_matrix.py:80]: Clever!

File: polyA/substitution_matrix.py [hunk 19]
```diff
@@ -99,8 +123,20 @@ def load_substitution_matrices(
         except StopIteration:
             break
 
+        # FIXME: will this break if there is no next line?
+        next_line = next(file)
```
Comment (glesica) [inline polyA/substitution_matrix.py:127]: Yep, use the trick above where you catch `StopIteration`, which is the exception that will be raised when the iterator is exhausted.

File: polyA/converters/alignment_to_stockholm.sh (hunk 12 above)
Comment (Audreym98) [reply polyA/converters/alignment_to_stockholm.sh:15]: Should move the CM converter too? So the converters directory (outside of the code dir) would have the RM converter, the CM converter, and this script

Comment (glesica) [reply polyA/converters/alignment_to_stockholm.sh]: Yeah, let's do that. I don't remember why the Stockholm converter ended up inside the module and we might as well have them all live together.

File: polyA/calculate_score.py [hunk 20]
```diff
@@ -30,11 +32,11 @@ def calculate_score(
...
 
     >>> sub_mat = {"AA":1, "AT":-1, "TA":-1, "TT":1}
-    >>> calculate_score(-5, -25, "AT", "AT", "", "", sub_mat)
+    >>> calculate_score(-5, -25, "AT", "AT", "", "", sub_mat, None)
     2.0
-    >>> calculate_score(-5, -25, "-T", "AT", "A", "A", sub_mat)
+    >>> calculate_score(-5, -25, "-T", "AT", "A", "A", sub_mat, None)
     -24.0
-    >>> calculate_score(-5, -25, "-T", "AT", "-", "", sub_mat)
+    >>> calculate_score(-5, -25, "-T", "AT", "-", "", sub_mat, None)
```
Comment (glesica) [inline polyA/calculate_score.py:39]: We should add a test for what happens when `char_complexity_adjustment` is not `None`.

File: polyA/_app.py (hunk 3 above)
Comment (Audreym98) [reply polyA/_app.py:99]: Whoops definitely didn't mean to open it twice

File: converters/alignment_to_stockholm.sh [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+alignment_file="$1"
+alignment_tool="$2"
+
+if [ "$alignment_tool" = "cross_match" ]; then
+  python cm_to_stockholm.py "$alignment_file"
+fi
+
+if [ "$alignment_tool" = "RepeatMasker" ]; then
+  python rm_to_stockholm.py "$alignment_file"
+fi
```
Comment (glesica) [inline converters/alignment_to_stockholm.sh:20]: We ought to account for this logic in the Python script instead of a shell script. Let's not worry about it here, but I suggest we create an issue for it.

File: polyA/constants.py [hunk 22]
```diff
@@ -54,6 +54,12 @@
 value doesn't matter, but it's here for clarity.
 """
 
+CROSS_MATCH_ADJUSTMENT = 0.999
+"""
+A constant value used in the equations from cross_match to calculate 
+the complexity adjusted score of an alignment.
+"""
```
Comment (glesica) [inline polyA/constants.py:61]: Why this particular value? Why not `0.9` or `0.9999`?

Comment (Audreym98) [reply polyA/constants.py:61]: I can't find a clear reason on why 0.999 is added, this is just the value given in the equations for how cross match performs complexity adjusted scoring

Comment (glesica) [reply polyA/constants.py:61]: We should explicitly state that it's just the value cross match uses, then. The current comment sort of does that, but it's not entirely clear, at least to me.
